zkevm-circuits/src/evm_circuit/util/memory_gadget.rs

```diff
@@ -686,3 +686,154 @@ impl<F: Field, const MAX_BYTES: usize, const ADDR_SIZE_IN_BYTES: usize>
+    {
+        fn configure_gadget_container(cb: &mut EVMConstraintBuilder<F>) -> Self {
+            let addr_start = cb.query_u64();
+            let addr_end = cb.query_u64();
+            let buffer_reader_gadget 
+                = BufferReaderGadget::<F, MAX_BYTES, ADDR_SIZE_IN_BYTES>::construct(cb, addr_start.expr(), addr_end.expr());
+            let bytes = cb.query_bytes();
+            let bytes_expr: Vec<Expression<F>> = bytes
+                .clone()
+                .iter()
+                .map(|e| e.expr())
+                .collect::<Vec<Expression<F>>>();
+            let buffer_reader_gadget_bytes_expr = (0..MAX_BYTES)
+                .map(|e| buffer_reader_gadget.byte(e))
+                .collect::<Vec<Expression<F>>>();
+            let buffer_reader_gadget_seletor = (0..MAX_BYTES)
+                .clone()
+                .map(|e| buffer_reader_gadget.read_flag(e))
+                .collect::<Vec<Expression<F>>>();
+
+            // test byte API
+            for (byte_expr, buffer_reader_gadget_byte_expr) in bytes_expr.iter().zip(buffer_reader_gadget_bytes_expr.iter()) {
+                cb.require_equal(
+                    "every byte equal",
+                    byte_expr.expr(),
+                    buffer_reader_gadget_byte_expr.expr(),
+                );
+            }
+
+            // test read_flag API
+            cb.require_equal(
+                "selector length equal",
+                sum::expr(buffer_reader_gadget_seletor),
+                addr_end.expr() - addr_start.expr(),
+            );
+
+            BufferReaderGadgetTestContainer { 
+                buffer_reader_gadget, 
+                addr_start,
+                addr_end, 
+                bytes,
+            }
+        }
+
+        fn assign_gadget_container(
+            &self,
+            witnesses: &[Word],
+            region: &mut CachedRegion<'_, '_, F>,
+        ) -> Result<(), Error> {
+            let offset = 0;
+            let addr_start= u64::from_le_bytes(witnesses[0].to_le_bytes()[..8].try_into().unwrap());
+            let addr_end= u64::from_le_bytes(witnesses[1].to_le_bytes()[..8].try_into().unwrap());
+            let mut input_bytes: Vec<u8> = Vec::new();
+            input_bytes.extend_from_slice(&witnesses[2].to_le_bytes());
+            self.addr_start.assign(region, offset, Some(addr_start.to_le_bytes()))?;
+            self.addr_end.assign(region, offset, Some(addr_end.to_le_bytes()))?;
+            self.bytes
+                .iter()
+                .zip(input_bytes.iter())
+                .map(|(byte, input_byte)| 
+                    byte.assign(region, offset, Value::known(F::from(*input_byte as u64)))
+                )
+                .collect::<Result<Vec<_>, _>>()?;
+            self.buffer_reader_gadget.assign(region, offset, addr_start, addr_end, &input_bytes[..])?;
+
+            Ok(())
+        }
+    }
+    
+    #[test]
```

Comment:
perhaps you could have a set of cases which has smaller `MAX_BYTES`, smaller than given buffer size or data size and to see how it goes.

Reply:
I can test MAX_BYTES < buffer size returning false but can't test against MAXBYTES < data size, because of this [assert](https://github.com/NOOMA-42/zkevm-circuits/blob/feat/%231513-BufferReaderGadget-unittest/zkevm-circuits/src/evm_circuit/util/memory_gadget.rs#L652) panic